File: Sources/XMTP/Messages/EncryptedPrivateKeyBundle.swift
//
//  EncryptedPrivateKeyBundle.swift
//
//
//  Created by Pat Nakajima on 11/17/22.
//

import XMTPProto

typealias EncryptedPrivateKeyBundle = Xmtp_MessageContents_EncryptedPrivateKeyBundle

extension EncryptedPrivateKeyBundle {
	func decrypted(with key: SigningKey) async throws -> PrivateKeyBundle {
		let signature = try await key.sign(message: Signature.enableIdentityText(key: v1.walletPreKey))
		let message = try Crypto.decrypt(signature.rawDataWithNormalizedRecovery, v1.ciphertext)

		return try PrivateKeyBundle(serializedData: message)
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/EncryptedPrivateKeyBundle.kt
package org.xmtp.android.library.messages

import kotlinx.coroutines.runBlocking
import org.xmtp.android.library.Crypto
import org.xmtp.android.library.PreEventCallback
import org.xmtp.android.library.SigningKey
import org.xmtp.android.library.XMTPException

typealias EncryptedPrivateKeyBundle = org.xmtp.proto.message.contents.PrivateKeyOuterClass.EncryptedPrivateKeyBundle

fun EncryptedPrivateKeyBundle.decrypted(
    key: SigningKey,
    preEnableIdentityCallback: PreEventCallback? = null,
): PrivateKeyBundle {
    preEnableIdentityCallback?.let {
        runBlocking {
            it.invoke()
        }
    }

    val signature = runBlocking {
        key.signLegacy(
            message = Signature.newBuilder().build()
                .enableIdentityText(key = v1.walletPreKey.toByteArray()),
        )
    } ?: throw XMTPException("Illegal signature")
    val message = Crypto.decrypt(signature.rawDataWithNormalizedRecovery, v1.ciphertext)
    return PrivateKeyBundle.parseFrom(message)
}
